docs/doc/90-contributing/01-rfcs/20230213-query-result-cache.md

```diff
@@ -0,0 +1,137 @@
+---
+title: Query Result Cache
+---
+
+- RFC PR: [datafuselabs/databend#10014](https://github.com/datafuselabs/databend/pull/10014)
+- Tracking Issue: [datafuselabs/databend#10011](https://github.com/datafuselabs/databend/issues/10011)
+
+## Summary
+
+Support query result cache for faster query response.
+
+## Motivation
+
+For some expensive queries with data that doesn't change frequently, we can cache the results to speed up the query response. For the same query with the same underlying data, we can return the cached result directly which improves the query efficiency greatly.
+
+For example, if we want to execute the following query to get the top 5 well-saled products everyt 10 seconds:
+
+```sql
+SELECT product, count(product) AS sales_count
+FROM sales_log
+GROUP BY product
+ORDER BY sales_count DESC
+LIMIT 5;
+```
+
+If we execute the full query pipeline every time, the cost may be every expensive, but the result is very small (5 rows). As the data of `sales_log` may not change very frequently, we can cache the result of the query and return the cached result directly for the same query.
+
+## Detail design
+
+### Result cache storage
+
+Databend uses key-value pairs to record query result caches. For every query, Databend will construct a key to represent the query, and store ralated information in the value.
+
+Databend will not store the query result directly in the key-value storage. Instead, Databend only stores the location of the result cache file in the value. The actual result cache will be stored in the storage layer (local fs, s3, ...).
+
+#### Key structure
+
+A query result cache key contains such information:
+
+- `ast`: the AST of the query.
+- `locations`: the locations of partitions used in the query.
+
+Databend will use the information to construct the key. The algorithm is like:
+
+```rust
+let ast_str = serialize(ast);
+let part_str = join(sort(locations), ",");
+let raw_key = join([ast_str, part_str], ",");
+let key = "_cache" + hash256(raw_key);
```

Comment:
Databend is a multi-tenant service. Do we need to add a tenant identifier in path here?